src/histogram.rs

```diff
@@ -257,31 +254,61 @@ impl Instant {
 
+            // It is different from `Instant::Monotonic`, the resolution here is millisecond.
+            // The processors in an SMP system do not start all at exactly the same time
+            // and therefore the timer registers are typically running at an offset.
+            // Use millisecond resolution for ignoring the error.
+            // See more: https://linux.die.net/man/2/clock_gettime
             #[cfg(all(feature="nightly", target_os="linux"))]
             Instant::MonotonicCoarse(t) => {
-                const NANOS_PER_SEC: f64 = 1_000_000_000.0;
-                let now = get_time(CLOCK_MONOTONIC_COARSE as clock_t);
-                if now.tv_sec > t.tv_sec || (now.tv_sec == t.tv_sec && now.tv_nsec >= t.tv_nsec) {
-                    Duration::new((t.tv_sec - now.tv_sec) as u64,
-                                  (t.tv_nsec - now.tv_nsec) as u32)
+                let now = get_time_coarse();
+                let now_ms = now.tv_sec * MILLIS_PER_SEC + now.tv_nsec / NANOS_PER_MILLI;
+                let t_ms = t.tv_sec * MILLIS_PER_SEC + t.tv_nsec / NANOS_PER_MILLI;
+                let dur = now_ms - t_ms;
+                if dur >= *CORASE_ERROR {
+                    Duration::from_millis(if dur > 0 { dur as u64 } else { 0 })
                 } else {
-                    panic!("system time jumped back, {:.9} -> {:.9}",
-                           t.tv_sec as f64 + t.tv_nsec as f64 / NANOS_PER_SEC,
-                           now.tv_sec as f64 + now.tv_nsec as f64 / NANOS_PER_SEC);
+                    panic!("system time jumped back, {:.3} -> {:.3}",
+                           t.tv_sec as f64 + t.tv_nsec as f64 / NANOS_PER_SEC as f64,
+                           now.tv_sec as f64 + now.tv_nsec as f64 / NANOS_PER_SEC as f64);
                 }
             }
         }
     }
 }
 
 #[cfg(all(feature="nightly", target_os="linux"))]
-fn get_time(clock: clock_t) -> timespec {
-    let mut t = timespec {
-        tv_sec: 0,
-        tv_nsec: 0,
-    };
-    assert_eq!(unsafe { clock_gettime(clock as _, &mut t) }, 0);
-    t
+use self::coarse::*;
+
+#[cfg(all(feature="nightly", target_os="linux"))]
+mod coarse {
+    use libc::{clock_gettime, clock_getres, CLOCK_MONOTONIC_COARSE};
+
+    pub use libc::timespec;
+
+    pub const NANOS_PER_SEC: f64 = 1_000_000_000.0;
+    pub const NANOS_PER_MILLI: i64 = 1_000_000;
+    pub const MILLIS_PER_SEC: i64 = 1_000;
+
+    pub fn get_time_coarse() -> timespec {
+        let mut t = timespec {
+            tv_sec: 0,
+            tv_nsec: 0,
+        };
+        assert_eq!(unsafe { clock_gettime(CLOCK_MONOTONIC_COARSE, &mut t) }, 0);
+        t
+    }
+
+    lazy_static! {
+        pub static ref CORASE_ERROR: i64 = {
+            let mut t = timespec {
+                tv_sec: 0,
+                tv_nsec: 0,
+            };
+            assert_eq!(unsafe { clock_getres(CLOCK_MONOTONIC_COARSE, &mut t) }, 0);
+            - t.tv_nsec / NANOS_PER_MILLI
```

Comment:
It is convenient for comparing, eg, error is -1, `2 >= error` and `-1 >= error` they are both true.